app/openai_utils.py

```diff
@@ -1,8 +1,14 @@
-import openai
+from openai import OpenAI
+import os
+
+def get_openai_client():
```

Comment:
I know you created this method to make it easier to mock the openai client. Is there a better way to do this?

Reply:
Certainly! We can use dependency injection to make it easier to mock the OpenAI client without needing to create a separate method. This approach allows us to pass the OpenAI client as a dependency to the functions that need it, making it more flexible and easier to test.

### Plan

1. **Refactor `generate_headline` to Accept OpenAI Client as a Parameter:**
   - Modify the `generate_headline` function to accept the OpenAI client as a parameter.

2. **Update the Test Cases:**
   - Update the test cases to pass the mocked OpenAI client to the `generate_headline` function.

### Steps

1. **Modify `app/openai_utils.py`:**
   - Refactor the `generate_headline` function to accept the OpenAI client as a parameter.

2. **Update `test_main.py`:**
   - Update the test cases to pass the mocked OpenAI client to the `generate_headline` function.

### Implementation

Let's start by refactoring the `app/openai_utils.py` file and updating the `test_main.py` file.

### Updated `app/openai_utils.py`

```python
from openai import OpenAI
import os

def generate_headline(client, text: str, image_url: str) -> str:
    prompt = f"Generate an ad headline for the following image URL based on the website text:\n\nWebsite Text:\n{text}\n\nImage URL: {image_url}\n\nAd Headline:"
    response = client.chat.completions.create(
        model="gpt-3.5-turbo",
        messages=[
            {"role": "system", "content": "You are a helpful assistant."},
            {"role": "user", "content": prompt}
        ],
        max_tokens=50
    )
    return response.choices[0].message['content'].strip()
```

### Updated `test_main.py`

```python
import pytest
from fastapi.testclient import TestClient
from unittest.mock import patch
import requests
from bs4 import BeautifulSoup
from app.main import app
from app.openai_utils import generate_headline
from app.scraping import extract_images
from test_utils import set_openai_api_key, mock_openai_client

client = TestClient(app)

def test_generate_headline():
    text = "This is a sample website text."
    image_url = "http://example.com/image.jpg"
    mock_client = mock_openai_client()
    headline = generate_headline(mock_client, text, image_url)
    assert headline == "Mocked Ad Headline"

def test_extract_text_success():
    url = "http://example.com"
    html_content = """
    <html>
        <body>
            <p>Hello, World!</p>
            <img src="http://example.com/image1.jpg" />
            <img src="http://example.com/image2.jpg" />
        </body>
    </html>
    """
    expected_text = "Hello, World!"
    expected_images = ["http://example.com/image1.jpg", "http://example.com/image2.jpg"]
    expected_headlines = ["Mocked Ad Headline", "Mocked Ad Headline"]

    with patch("requests.get") as mock_get, patch("app.openai_utils.OpenAI", return_value=mock_openai_client()):
        mock_get.return_value.status_code = 200
        mock_get.return_value.content = html_content

        response = client.get("/extract-text", params={"url": url})
        assert response.status_code == 200
        assert response.json() == {"text": expected_text, "images": expected_images, "headlines": expected_headlines}

def test_extract_text_invalid_url():
    response = client.get("/extract-text", params={"url": "invalid-url"})
    assert response.status_code == 422

def test_extract_text_request_exception():
    url = "http://example.com"

    with patch("requests.get") as mock_get:
        mock_get.side_effect = requests.RequestException("Request failed")

        response = client.get("/extract-text", params={"url": url})
        assert response.status_code == 400
        assert response.json() == {"detail": "Request failed"}

def test_extract_text_bermuda():
    url = "https://thinkingofbermuda.com"
    
    with patch("app.openai_utils.OpenAI", return_value=mock_openai_client()):
        response = client.get("/extract-text", params={"url": url})
        assert response.status_code == 200
        assert "text" in response.json()
        assert "images" in response.json()
        assert "headlines" in response.json()

def test_extract_images():
    html_content = """
    <html>
        <body>
            <img src="http://example.com/image1.jpg" />
            <meta property="og:image" content="http://example.com/image2.jpg" />
            <link rel="image_src" href="http://example.com/image3.jpg" type="image/jpeg" />
            <source srcset="http://example.com/image4.jpg 1x, http://example.com/image5.jpg 2x" />
        </body>
    </html>
    """
    soup = BeautifulSoup(html_content, "html.parser")
    expected_images = [
        "http://example.com/image1.jpg",
        "http://example.com/image2.jpg",
        "http://example.com/image3.jpg",
        "http://example.com/image4.jpg"
    ]
    assert extract_images(soup) == expected_images
```

### Explanation of Changes

1. **Refactored `generate_headline`:**
   - Modified the `generate_headline` function to accept the OpenAI client as a parameter.

2. **Updated Test Cases:**
   - Updated the test cases to pass the mocked OpenAI client to the `generate_headline` function.

I will now update the PR with these changes.